document-legacy/src/document.rs

```diff
@@ -559,50 +621,156 @@ impl Document {
+					aggregate_insertions(folder, &mut destination_path.as_slice().to_vec(), &mut responses, duplicating);
 				};
 
-				responses.extend([DocumentChanged, CreatedLayer { path: destination_path.clone() }, FolderChanged { path: folder_path.to_vec() }]);
-				responses.extend(update_thumbnails_upstream(&destination_path));
 				Some(responses)
 			}
 			Operation::DuplicateLayer { path } => {
+				// Notes for review: I wasn't sure on how to apply unwrap_or() to lines of code that use the funciton self.layer()
+
 				let layer = self.layer(&path)?.clone();
 				let (folder_path, _) = split_path(path.as_slice()).unwrap_or((&[], 0));
-				let folder = self.folder_mut(folder_path)?;
 
+				let layer_is_folder = layer.as_folder().is_ok();
+
+				// Recursively collect each of the nested folders and shapes if the layer is a folder
+				let mut duplicated_layers = if layer_is_folder { recursive_collect(self, &path) } else { Vec::new() };
+				// Duplicate the Layer objects for our duplicated layers
+				// Sort both vectors by the layer path size, from shallowest to deepest
+				let duplicated_layers_objects: Vec<Layer> = duplicated_layers.iter().map(|p| self.layer(p.as_slice()).unwrap()).cloned().collect();
+				let mut indices: Vec<usize> = (0..duplicated_layers.len()).collect();
+				indices.sort_by_key(|&i| duplicated_layers[i].len());
+				duplicated_layers.sort_by_key(|a| a.len());
+				let duplicate_layer_objects_sorted: Vec<&Layer> = indices.iter().map(|&i| &duplicated_layers_objects[i]).collect();
+
+				let folder = self.folder_mut(folder_path)?;
 				let selected_id = path.last().copied().unwrap_or_default();
 				let insert_index = folder.layer_ids.iter().position(|&id| id == selected_id).unwrap_or(0) as isize + 1;
 
 				if let Some(new_layer_id) = folder.add_layer(layer, None, insert_index) {
-					let new_path = [folder_path, &[new_layer_id]].concat();
+					let new_path: &Vec<u64> = &([folder_path, &[new_layer_id]].concat());
+
+					let mut responses: Vec<DocumentResponse> = vec![];
+					responses.extend([
+						DocumentChanged,
+						CreatedLayer {
+							path: new_path.to_vec(),
+							is_selected: true,
+						},
+						FolderChanged { path: folder_path.to_vec() },
+					]);
+					responses.extend(update_thumbnails_upstream(path.as_slice()));
+
+					if layer_is_folder {
+						let mut old_to_new_layer_id: HashMap<LayerId, LayerId> = HashMap::new();
+						let new_folder = self.folder_mut(new_path)?;
+						// Clear the new folders layer_ids/layers because they contain the layer_ids/layers of the layer were duplicated
+						// Generate a new next assignment ID to avoid collision
+						new_folder.layer_ids = vec![];
+						new_folder.layers = vec![];
+						new_folder.generate_new_folder_ids();
+
+						for (counter, duplicate_layer) in duplicated_layers.into_iter().enumerate() {
+							let old_layer_id = *duplicate_layer.last().unwrap_or(&0);
```

Comment:
is unwrap_or the move here? is this variable any good to us if its 0? maybe an "if let some old_layer_id"

Reply:
fixed